src/control/lib/control/storage.go

```diff
@@ -34,19 +34,19 @@ var storageHashOpts = hashstructure.HashOptions{
 type HostStorage struct {
 	// NvmeDevices contains the set of NVMe controllers (SSDs)
 	// in this configuration.
-	NvmeDevices storage.NvmeControllers `json:"nvme_devices"`
+	NvmeDevices *storage.NvmeControllers `json:"nvme_devices"`
```

Comment:
yes but hashstructure doesn't distinguish between a typed nil slice and an empty slice when generating a hash digest, so in order to generate distinct keys in the HSM I needed to change the field to a pointer, in which case a nil pointer could be distinguished from a pointer to an empty slice (for failure and success respectively) which would each generate a distinct key for the map when hashed

Reply:
This is the crux of the issue I am having with this PR. Because a third-party library doesn't behave the way we want it to in order to handle this narrow use case, we're making changes all throughout our codebase? What if the next version fixes it, or otherwise changes behavior? Will we have more ripples throughout our code to accommodate that?

Given that the code is open source under a permissive license, I think that we could fork it, fix it, use it and submit the fix upstream. I suspect that they would accept a contribution to fix this because after studying the code I think it's an oversight rather than a deliberate design decision.

Alternatively, we could use this as an opportunity to rethink what we're doing here. I'm still not convinced that we need to perform remote NVMe prepare/reset. The server already does an automatic prepare on startup. This seems to me like we're investing a lot of time and energy on solving a problem with limited application.